source/blog/2021-05-24-orm2.9.md

```diff
@@ -0,0 +1,163 @@
+
+Example:
+
+```php
+use Doctrine\ORM\Mapping AS ORM;
+
+#[ORM\Entity(repositoryClass: UserRepository::class)]
+class User
+{
+    #[ORM\Id, ORM\Column, ORM\GeneratedValue]
+    public ?int $id = null;
+
+    #[ORM\Column]
+    public \DateTime $created;
+
+    #[ORM\ManyToOne]
+    public Email $email;
+}
+```
+
+## Psalmified APIs
+
+Improved the documentation to make sure static analysis tools and IDEs know
+about the right entity classes returned from `EntityManager`,
+`EntityRepository` and other public ORM APIs. This includes generics support
+when you extend `EntityRepository`.
+
+```php
+use Doctrine\ORM\EntityRepository;
+use App\Entity\User;
+
+/**
+ * @template-extends EntityRepository<User>
+ */
+class UserRepository extends EntityRepository
+{
+}
+```
+
+## Query::HINT_READ_ONLY
+
+A new query hint is added that allows hydrating entities through DQL that are
+marked as read only for the unit of work session, as long as they are not yet
+loaded as writeable:
+
+```php
+$dql = 'SELECT u FROM ' . ReadOnlyEntity::class . ' u WHERE u.id = ?1';
+
+$query = $entityManager->createQuery($dql);
+$query->setParameter(1, $user->id);
+$query->setHint(Query::HINT_READ_ONLY, true);
+
+$user = $query->getSingleResult();
+```
+
+## Index/UniqueConstraints using Field Names
+
+Instead of specifying column names for an index or unique-constraint declaration
+you can now alternatively use field names.
+
+```php
+
+use Doctrine\ORM\Mapping AS ORM;
+
+#[ORM\Entity]
+#[ORM\Index(fields: ["isPublished"])]
+class Post
+{
+    #[ORM\Column]
+    public bool $isPublished = false;
+}
+```
+
+This simplifies mapping as the column names passed through the naming strategy
+do not need to be known.
+
+## INDEX BY Associations
+
+Previously DQL `INDEX BY` was not possible for assocations, now you can:
+
+```php
+$dql = 'SELECT p, u FROM Post INDEX BY p.author JOIN p.author u WHERE p.id = 3';
+```
+
+## Deprecations
+
+Doctrine ORM 2.9 rethinks deprecations and integrates with our new
+[doctrine/deprecations](https://github.com/doctrine/deprecations/) library.
+
+- Undeprecate `merge()` and `detach()` as no replacements are available yet
+- Notify Persist Change Tracking: Use Explicit Change Tracking instead
+- DQL `SELECT PARTIAL` syntax, use Value Objects with `SELECT NEW` instead
+- `EntityManager::flush()` with arguments
+- `EntityManager::clear()` with arguments (use detach)
+- Named Queries in Mapping (use Repository)
+- `cli-config.php` for console command configuration, inject `EntityManagerProvider` instead.
+- Deprecate `doctrine/cache` for metadata caching, use PSR-6 cache instead
+
+## PHP 7.1 Support
+
+ORM 2.9 reintroduces PHP 7.1 support, because it wasn't technically unsupported
+anyways.  No changes were necessary to the code to allow it again except in the
+testsuite.
+
+The PHP 7.1 support was re-added to allow a very broad approach to prepare for
+some of the deprecations that we are going to introduce in ORM 3.
```

Comment:
Are you talking about code that will be deprecated in 3.0, or deprecated in 2.x and dropped in 3.0?

Reply:
good point!